File: desktop/engine/vad.py
"""Voice Activity Detection — only checks the tail of audio."""
import numpy as np

SAMPLE_RATE = 16000


class VAD:
    """Detects speech end via RMS energy on the tail only."""

    def __init__(self, silence_threshold=0.001, silence_duration=8.0, sr=SAMPLE_RATE):
        self.silence_threshold = silence_threshold  # very low — some mics are quiet
        self.silence_duration = silence_duration
        self.sr = sr

    def is_silence(self, audio_chunk):
        if len(audio_chunk) == 0:
            return True
        rms = np.sqrt(np.mean(audio_chunk ** 2))
        return rms < self.silence_threshold

    def detect_end_of_speech(self, full_audio):
        """Check if audio TAIL has enough silence. Only reads last N seconds."""
        silence_samples = int(self.silence_duration * self.sr)
        if len(full_audio) < silence_samples:
            return False

        # Only look at the tail — not the whole buffer
        tail = full_audio[-silence_samples:]
        window = int(0.5 * self.sr)  # 500ms windows (fewer checks)
        for i in range(0, len(tail), window):
            chunk = tail[i:i + window]
            if not self.is_silence(chunk):
                return False
        return True

    def has_speech(self, full_audio, min_speech_duration=0.3):
        """Check if audio contains any speech. Only scans first 10s + last 10s."""
        window = int(0.2 * self.sr)
        needed = int(min_speech_duration / 0.2)
        speech_windows = 0

        # Only scan first 10s and last 10s (not entire buffer)
        scan_samples = int(10 * self.sr)
        if len(full_audio) > scan_samples * 2:
            regions = [full_audio[:scan_samples], full_audio[-scan_samples:]]
        else:
            regions = [full_audio]

        for region in regions:
            for i in range(0, len(region), window):
                chunk = region[i:i + window]
                if not self.is_silence(chunk):
                    speech_windows += 1
                if speech_windows >= needed:
                    return True
        return False

File: rs/src/vad.rs
/// Voice Activity Detection — RMS-based, only checks tail of audio.
pub struct Vad {
    silence_threshold: f32,
    silence_duration: f32,
    sample_rate: u32,
}

impl Vad {
    pub fn new(silence_duration: f32, sample_rate: u32) -> Self {
        Self {
            silence_threshold: 0.005,
            silence_duration,
            sample_rate,
        }
    }

    fn rms(chunk: &[f32]) -> f32 {
        if chunk.is_empty() {
            return 0.0;
        }
        let sum: f32 = chunk.iter().map(|s| s * s).sum();
        (sum / chunk.len() as f32).sqrt()
    }

    fn is_silence(&self, chunk: &[f32]) -> bool {
        Self::rms(chunk) < self.silence_threshold
    }

    /// Check if audio tail has enough silence to stop.
    pub fn detect_end_of_speech(&self, audio: &[f32]) -> bool {
        let silence_samples = (self.silence_duration * self.sample_rate as f32) as usize;
        if audio.len() < silence_samples {
            return false;
        }

        let tail = &audio[audio.len() - silence_samples..];
        let window = (0.5 * self.sample_rate as f32) as usize;

        for chunk in tail.chunks(window) {
            if !self.is_silence(chunk) {
                return false;
            }
        }
        true
    }

    /// Check if audio contains any speech.
    pub fn has_speech(&self, audio: &[f32]) -> bool {
        let window = (0.2 * self.sample_rate as f32) as usize;
        let needed = 2; // ~0.4s of speech
        let mut speech_windows = 0;

        // Only scan first 10s and last 10s
        let scan = (10.0 * self.sample_rate as f32) as usize;
        let regions: Vec<&[f32]> = if audio.len() > scan * 2 {
            vec![&audio[..scan], &audio[audio.len() - scan..]]
        } else {
            vec![audio]
        };

        for region in regions {
            for chunk in region.chunks(window) {
                if !self.is_silence(chunk) {
                    speech_windows += 1;
                    if speech_windows >= needed {
                        return true;
                    }
                }
            }
        }
        false
    }
}
